src/wasmedge/wasmedge.cc

```diff
@@ -0,0 +1,600 @@
+  };
+  FOR_ALL_WASM_VM_IMPORTS(_REGISTER_HOST_FUNCTION)
+#undef _REGISTER_HOST_FUNCTION
+
+#define _GET_MODULE_FUNCTION(T)                                                                    \
+  void getFunction(std::string_view function_name, T *f) override {                                \
+    getModuleFunctionImpl(function_name, f);                                                       \
+  };
+  FOR_ALL_WASM_VM_EXPORTS(_GET_MODULE_FUNCTION)
+#undef _GET_MODULE_FUNCTION
+private:
+  template <typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                void (*function)(Args...));
+
+  template <typename R, typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                R (*function)(Args...));
+
+  template <typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<void(ContextBase *, Args...)> *function);
+
+  template <typename R, typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<R(ContextBase *, Args...)> *function);
+
+  WasmEdgeLoaderPtr loader_;
+  WasmEdgeValidatorPtr validator_;
+  WasmEdgeExecutorPtr executor_;
+  WasmEdgeStorePtr store_;
+  WasmEdgeASTModulePtr module_;
+  WasmEdge_MemoryInstanceContext *memory_;
+
+  std::unordered_map<std::string, HostFuncDataPtr> host_functions_;
+  std::unordered_map<std::string, HostModuleDataPtr> import_objects_;
+  std::unordered_set<std::string> module_functions_;
+};
+
+bool WasmEdge::load(std::string_view bytecode, std::string_view,
+                    const std::unordered_map<uint32_t, std::string>) {
+  WasmEdge_ASTModuleContext *mod = nullptr;
+  WasmEdge_Result res = WasmEdge_LoaderParseFromBuffer(
+      loader_.get(), &mod, reinterpret_cast<const uint8_t *>(bytecode.data()), bytecode.size());
+  if (!WasmEdge_ResultOK(res)) {
+    return false;
+  }
+  res = WasmEdge_ValidatorValidate(validator_.get(), mod);
+  if (!WasmEdge_ResultOK(res)) {
+    return false;
+  }
+  module_ = mod;
+  return true;
+}
+
+bool WasmEdge::link(std::string_view debug_name) {
+  assert(module_ != nullptr);
+
+  // Create store and register imports.
+  store_ = WasmEdge_StoreCreate();
+  if (store_ == nullptr) {
+    return false;
+  }
+  for (auto &&it : import_objects_) {
+    auto res = WasmEdge_ExecutorRegisterImport(executor_.get(), store_.get(), it.second->cxt_);
+    if (!WasmEdge_ResultOK(res)) {
+      fail(FailState::UnableToInitializeCode,
+           std::string("Failed to link Wasm module due to import: ") + it.first);
+      return false;
+    }
+  }
+  // Instantiate module.
+  if (auto res = WasmEdge_ExecutorInstantiate(executor_.get(), store_.get(), module_.get());
+      !WasmEdge_ResultOK(res)) {
```

Comment:
This is pretty much Go-style, and not C++. It's technically valid, but not very readable for C++ developers.

Could you split this into separate statements?
```
auto res = WasmEdge_ExecutorInstantiate(executor_.get(), store_.get(), module_.get());
if (!WasmEdge_ResultOK(res)) {
  ..
}
```

Reply:
Fixed.